### PR TITLE
Don't attempt to serialize empty response data.

### PR DESCRIPTION
**Pull request by paulyoung**
This addresses the issue described in #31.

### CODE REVIEW

**Comment by paulyoung** [comment]
I added a test and some fakes but there's probably a better way to do that.

File: Tests/Fakes/FakeRequest.swift [hunk 1]
```diff
@@ -30,3 +30,19 @@ struct FakeNullDataRequest: Request {
...
+  }
+  
+  func parse(j: JSON) -> Result<EmptyResponse, NSError> {
+    switch j {
+    case JSON.Null: return Result.Success(())
+    default:
+      return Result.Failure(Result<EmptyResponse, NSError>.error())
+    }
+  }
+}
```
Comment (paulyoung) [inline Tests/Fakes/FakeRequest.swift:48]: Not sure what I was thinking here, this is the same as `FakeNullDataRequest`.

File: Tests/Tests/APIClientSpec.swift [hunk 2]
```diff
@@ -66,6 +66,20 @@ class APIClientSpec: QuickSpec {
...
+              it("returns a null JSON object") {
+                let request = FakeEmptyDataRequest()
+                let performer = FakeEmptyResponseRequestPerformer()
+                
+                let client = APIClient(requestPerformer: performer)
+                var result: Result<EmptyResponse, NSError>?
+                
+                client.performRequest(request) { result = $0 }
+                
+                expect(result!.value!).toEventually(beVoid())
```
Comment (paulyoung) [inline Tests/Tests/APIClientSpec.swift:80]: Without the `where` clause I added, this will crash. Not ideal.

**Comment by paulyoung** [comment]
@tonyd256 @gfontenot any suggestion on how to improve this?

**Comment by paulyoung** [comment]
Probably going to take another look at this in the morning.

**Comment by sidraval** [comment]
LGTM :+1:

File: Tests/Fakes/FakeRequest.swift (hunk 1 above)
Comment (paulyoung) [inline Tests/Fakes/FakeRequest.swift:48]: I guess I was trying to distinguish between null data and data with zero bytes, even though the implementations are currently the same.

**Comment by paulyoung** [comment]
Thanks @sidraval. It definitely still feels half baked on the testing side to me.

**Comment by sidraval** [comment]
@paulyoung Perhaps I'm missing something, can you clarify what feels half baked?

**Comment by paulyoung** [comment]
I think mainly:
1. There was already a test for null data, but not data with 0 bytes which is the crux of this issue. I added `FakeEmptyDataRequest` to correspond to `FakeNullDataRequest` but they have the same implementation. I feel like maybe they shouldn't and I may have overlooked something there.
2. Similar story for `FakeEmptyResponseRequestPerformer`. It's close to `FakeRequestPerformer` so maybe they could be combined, or it's a sign that some things can be removed from `FakeEmptyResponseRequestPerformer`.
3. The test case I added will crash if there's a regression. I should fix that.

**Comment by paulyoung** [comment]
Although for number 3, the other tests do the same...

**Comment by gfontenot** [comment]
I'm :+1: on this change in general. I'm also OK with merging these tests as they are, and then trying to refactor these fakes a bit later to clean it up. Fixing this issue was in my mental TODO list (that would _really_ be better not in my head), so I think it's a really good change.

**Comment by gfontenot** [comment]
So, basically, @paulyoung if you feel like 80% good with this as it stands, we can merge it and come back to do a refactoring pass later.

**Comment by paulyoung** [comment]
Sounds good!

**Comment by paulyoung** [comment]
@gfontenot good to go? :bow: :pray:

**Comment by gfontenot** [comment]
@paulyoung yup yup! Merged in as 5911762. Thanks!

**Pull request closed without merging**